Allow limiting the number of outstanding prefetch requests

kschoelhorn opened the pull request:
Currently prefetch will send all requests for all chunks for the file in one shot. With the default chunk size of 32k, this can result in many thousand outstanding requests for large files. Some servers like Serv-U 15.2.3.742 seem to be dropping requests after a certain number, which results in the file download hanging indefinitely (or until the server closes the connection).

Fix this issue by letting the user specify a limit for the number of outstanding requests. This is similar to openssh's sftp, which limits the number of outstanding requests to 64 by default.

Note that this is just a quick initial implementation to fix the problem I had while using paramiko with Serv-U. This could also be a setting of the SFTPClient, that limits any `_async_request`s, not just those from prefetching, though it seems the only other user is `listdir_iter`, which already has the `read_aheads` parameter, which also limits the number of requests.

See also #1779 where this was previously proposed and #151, which mixes different issues, but also contains descriptions of get calls hanging, that are fixed by disabling prefetching and are probably the same issue.

bskinn commented:
Looks like this is specific to SFTP?

bskinn commented on paramiko/sftp_file.py:
```diff
@@ -522,18 +530,26 @@ def _get_size(self):
+            if max_outstanding_requests is not None:
+                while True:
+                    with self._prefetch_lock:
+                        if len(self._prefetch_extents) < max_outstanding_requests:
```
Checking my understanding of the implementation here -- does each call to `self.sftp._async_request()` cause `self._prefetch_extents` to increment by one, and then once each chunk is received, `._prefetch_extents` is decremented?

And, thus, this new code causes this queuing within `self._prefetch_thread()` to idle whenever we hit our max-requests cap?

kschoelhorn replied:
The new async request number returned by `_async_request` is added to `_prefetch_extents` by the code below (hidden by the diff view on github by default:

```
            with self._prefetch_lock:
                self._prefetch_extents[num] = (offset, length)
```

The request number is then removed again in `_async_response`.

So yes, the background thread just idles here if the limit has been reached.

bskinn commented on paramiko/sftp_file.py:
```diff
@@ -473,9 +477,9 @@ def prefetch(self, file_size=None):
+            self._start_prefetch(chunks, max_outstanding_requests)
 
-    def readv(self, chunks):
+    def readv(self, chunks, max_outstanding_requests = None):
```
It looks like you haven't added any use of `max_outstanding_requests` in `readv()` here -- seems like either it shouldn't be added, or there's a piece of implementation missing.

kschoelhorn replied:
Good catch! I meant to forward the parameter to `_start_prefetch`.

bskinn reviewed (changes requested):
Might be a piece of implementation missing

bskinn commented:
Will need bitprophet's input on that parameter name, I think -- paramiko tends toward more concise names.

bskinn commented:
@kschoelhorn Are there any tests that could be added for this, that would be less onerous than mocking a server and monitoring the length of the request queue?

bskinn commented:
Per https://github.com/paramiko/paramiko/issues/1587#issuecomment-657647904 and the [SO post](https://stackoverflow.com/a/58339631) linked therein, this feature might mitigate or solve #1587?

The workaround posted in [this Gist](https://gist.github.com/vznncv/cb454c21d901438cc228916fbe6f070f), linked from https://github.com/paramiko/paramiko/issues/1587#issuecomment-1137274920, suggests it likely would help.

The implementation in this PR adds a parameter to the public API, versus the private class/member implementation in the Gist, so this PR is likely preferred.

kschoelhorn commented:
> @kschoelhorn Are there any tests that could be added for this, that would be less onerous than mocking a server and monitoring the length of the request queue?

I'll think about some tests. I think we should be able to mock `SFTPClient` and then simulate a large file transfer.

bwinston-sdp commented:
Is this still a possible PR pending tests? I've run into a similar situation as mentioned in the issues/Gist linked, and this works beautifully. I'm happy to rebase/write some tests if @kschoelhorn is MIA and/or wants to pass it off.

bskinn commented:
> Is this still a possible PR pending tests?

Yes, it definitely is -- it's not at the top of the priority list for paramiko at the moment, though, so it hasn't seen any specific action.

Up to you and @kschoelhorn to negotiate the test work. Whoever would like to contribute them is welcome!

bwinston-sdp commented:
Not sure if it was the "right way" to do it, but I created a PR to @kschoelhorn's branch, rebased on main, with lint fixes and an added test. LMK if I should do it another way (create a new PR directly to paramiko with changes, reference/credit this one?).

kschoelhorn commented:
@bwinston-sdp: Thank you very much for the test, I'll check and merge to this PR later!

bwinston-sdp commented:
Thanks @kschoelhorn!! I figured the test might be flaky because of the background thread, but wasn't sure exactly how to ensure stability (I've not done much multi-threaded coding).

One other thing I thought of -- I know the `max_outstanding_requests` gets passed through `readv` as well, not sure if we need another test for that.

kschoelhorn commented:
> One other thing I thought of -- I know the `max_outstanding_requests` gets passed through `readv` as well, not sure if we need another test for that.

Both use the same prefetch code internally, so I think one test is enough.

bskinn commented on paramiko/sftp_client.py:
```diff
@@ -773,6 +780,10 @@ def getfo(self, remotepath, fl, callback=None, prefetch=True):
         :param bool prefetch:
             controls whether prefetching is performed (default: True)
+        :param int max_outstanding_prefetch_requests:
+            Limit the number of outstanding requests instead of sending all
```
Is this parameter defining the maximum _outstanding_ requests, or the max _concurrent_ requests? It seems more the latter to me.

Separately, this parameter is an `int` -- its description should explicitly say how that value is used. It's _implicit_ that the value is the max number of concurrent requests, but it should be made explicit.

kschoelhorn replied:
I meant outstanding in the sense of "not yet done": the requests were sent, but we have not yet received an answer. But concurrent is also fine! :slightly_smiling_face:

bskinn replied:
@kschoelhorn Yeah, that was what prompted my suggestion: to my mind, 'outstanding' could be ambiguous, meaning either:

1. The number of requests currently in-flight (the intended meaning)
2. The total number of requests yet to be completed, including both in-flight and not-yet-sent requests (not the intended meaning)

Meaning (2) doesn't make a whole lot of sense, practically, but if that's the sense of the word that someone reaches first, then the name could be confusing. Thus, my recommendation for 'concurrent'.

bskinn commented:
Also--please add a CHANGELOG entry to `/sites/www/changelog.rst` -- thanks!

bwinston-sdp commented:
I updated these and pushed to the PR on @kschoelhorn's branch.

bskinn commented:
:+1: - Linking directly to https://github.com/kschoelhorn/paramiko/pull/1, for reference.

kschoelhorn commented:
Thanks again to @bwinston-sdp! I merged your changes, rebased to latest master and fixed a few lints.

@bskinn: I guess you can remove the "Needs changelog/docs" flag now. 🙂

bskinn commented:
> I guess you can remove the "Needs changelog/docs" flag now. 🙂

Done!

bitprophet commented on paramiko/sftp_file.py:
```diff
@@ -485,6 +492,13 @@ def readv(self, chunks):
             a list of ``(offset, length)`` tuples indicating which sections of
             the file to read
+        :param int max_concurrent_prefetch_requests:
+            The maximum number of concurrent read requests to prefetch.
```
I'd replace this param description everywhere but `SFTPClient.get`, with something like:

```
The maximum number of concurrent read requests to prefetch. See the API docs for `.SFTPClient.get` for details.
```

(IOW I don't think it's useful to have the same paragraph in all 4+ places :))

TallGibbs commented:
Sorry if this seems really silly, but all this testing and (proposed) change seems like exactly what I need to fix my current issue in some SFTP code.  Did this change get implemented?  Is there documentation for how to specify the "max_outstanding_prefetch_requests" variable?  I am trying to transfer 1GB file through SFTP and the host is closing the connection with the "EOFError"

bwinston-sdp commented:
@TallGibbs this feature is included in 3.3, with an additional parameter in the `get()` function: https://docs.paramiko.org/en/latest/api/sftp.html?highlight=prefetch#paramiko.sftp_client.SFTPClient.get

TallGibbs commented:
> @TallGibbs this feature is included in 3.3, with an additional parameter in the `get()` function: https://docs.paramiko.org/en/latest/api/sftp.html?highlight=prefetch#paramiko.sftp_client.SFTPClient.get

Oh awesome!  Looks like I need to update then....I'm currently using 2.8.1

Thank you for verifying

TallGibbs commented:
@bwinston-sdp I just updated to 3.4.0 and modified my code to include get(file, local_path, max_concurrent_prefetch_requests=32) and it worked like a charm!  Super simple implementation (WAY easier than all the solutions on S.O.).  This was a file just under 1 GB, for anyone else encountering these issues.

Thanks to everyone involved for solving this SFTP problem
